lib/src/metta/runner/modules/catalog.rs

```diff
@@ -0,0 +1,568 @@
+//!      ╱                    ╱       ⎽⎼⎻⎺  pkg-info in  ⎺⎻⎼⎽ Yes      ⎽⎼⎻⎺pkg-info entry ⎺⎻⎼⎽ No
+//!     ╱  (import! module)  ╱─────►<   &self has entry for   >─────►<   has fs_path attrib?   >───┐
+//!    ╱                    ╱         ⎺⎻⎼⎽    module?    ⎽⎼⎻⎺          ⎺⎻⎼⎽               ⎽⎼⎻⎺     │
+//!   └────────────────────┘               ⎺⎻⎼⎽ ⎽⎽⎽ ⎽⎼⎻⎺                    ⎺⎻⎼⎽ ⎽⎽⎽ ⎽⎼⎻⎺          │
+//!                                              │ No                             │ Yes            │
+//!  ┌─────────────────────────┐     ┌───────────▼─────────────┐      /───────────▼─────────────\  │
+//!  │  Query ModuleCatalogs   │     │    Assume any module    │      │    Load the module at   │  │
+//!  │     in order, with      │◄──┬─┤  version will satisfy   │      │   the file-system path  │  │
+//!  │   version requirement   │   │ │       dependency        │      │   with first successful │  │
+//!  │                         │   │ │                         │      │      FsModuleFormat     │  │
+//!  └───────────┬─────────────┘   │ └───────────▲─────────────┘      \───────────▲─────────────/  │
+//!              │                 │             │                                │                │
+//!  /───────────▼─────────────\   │             │                    ┌───────────┴─────────────┐  │
+//!  │  Load the module from   │   │             │                    │    clone module from    │  │
+//!  │   the first catalog     │   │             │                    │     remote repo to      │  │
+//!  │     that reports a      │   │             │                    │    local resource dir   │  │
+//!  │    successful match     │   │             │                    │                         │  │
+//!  \─────────────────────────/   │             │                    └───────────▲─────────────┘  │
+//!                                │             │ No                             │ Yes            │
+//!                                │Yes    ⎽⎼⎻⎺ ⎺⎺⎺ ⎺⎻⎼⎽                    ⎽⎼⎻⎺ ⎺⎺⎺ ⎺⎻⎼⎽          │
+//!                                │  ⎽⎼⎻⎺pkg-info entry ⎺⎻⎼⎽       No ⎽⎼⎻⎺pkg-info entry ⎺⎻⎼⎽     │
+//!                                └<   has version attrib?   >◄─────<     has git attrib?     >───┘
+//!                                   ⎺⎻⎼⎽               ⎽⎼⎻⎺          ⎺⎻⎼⎽               ⎽⎼⎻⎺
+//!                                        ⎺⎻⎼⎽ ⎽⎽⎽ ⎽⎼⎻⎺                    ⎺⎻⎼⎽ ⎽⎽⎽ ⎽⎼⎻⎺
+//! ```
+//!
+
+//LP-TODO-NEXT make a test to make sure circular imports are caught and don't lead to infinite recursion
+//QUESTION: Should circular imports between modules be allowed?  The current implementation (and possibly
+// the MeTTa language itself) disallow circular imports because there is no concept of forward declaration.
+// It *may* be possible to pre-parse the language in order to make recursive imports possible, but I have
+// not yet thought in detail about this.
```

Comment:
The current state of the code isn't great because loading a module on-demand can still cause problems if a sub-module is referenced by a module loader, before the module has finished initialization.

I think I need to spend a few hours on this, to make sure I have tested all the edge cases, and write unit tests.  But it's not a block-merge.